File: packages/mp_text/ios/Classes/MpTextPlugin.swift
import Flutter
import MediaPipeTasksText
import UIKit

/// iOS bridge for MediaPipe's mobile-only text generation tasks.
public final class MpTextPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
  private static let methodChannelName = "dev.ifiokjr.mp_text/methods"
  private static let eventChannelName = "dev.ifiokjr.mp_text/events"

  private let worker = DispatchQueue(
    label: "dev.ifiokjr.mp_text.worker",
    qos: .userInitiated,
    attributes: .concurrent
  )
  private let lock = NSLock()
  private var nextHandle: Int64 = 1
  private var proofreaders: [Int64: TextProofreader] = [:]
  private var summarizers: [Int64: TextSummarizer] = [:]
  private var eventSink: FlutterEventSink?
  private var methodChannel: FlutterMethodChannel?
  private var eventChannel: FlutterEventChannel?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let methodChannel = FlutterMethodChannel(
      name: methodChannelName,
      binaryMessenger: registrar.messenger()
    )
    let eventChannel = FlutterEventChannel(
      name: eventChannelName,
      binaryMessenger: registrar.messenger()
    )
    let instance = MpTextPlugin()
    instance.methodChannel = methodChannel
    instance.eventChannel = eventChannel
    registrar.addMethodCallDelegate(instance, channel: methodChannel)
    eventChannel.setStreamHandler(instance)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    do {
      switch call.method {
      case "proofreader.create":
        try createProofreader(call, result: result)
      case "proofreader.proofread":
        try proofread(call, result: result)
      case "proofreader.stream":
        try proofreadStreaming(call, result: result)
      case "proofreader.close":
        try closeProofreader(call, result: result)
      case "summarizer.create":
        try createSummarizer(call, result: result)
      case "summarizer.summarize":
        try summarize(call, result: result)
      case "summarizer.stream":
        try summarizeStreaming(call, result: result)
      case "summarizer.close":
        try closeSummarizer(call, result: result)
      default:
        result(FlutterMethodNotImplemented)
      }
    } catch {
      result(flutterError(error, code: "invalid_argument"))
    }
  }

  private func createProofreader(_ call: FlutterMethodCall, result: @escaping FlutterResult) throws
  {
    let arguments = try call.argumentsMap()
    let modelPath = try arguments.requiredString("modelPath")
    let maxTokens = arguments.optionalInt("maxTokens")
    worker.async { [weak self] in
      guard let self else { return }
      do {
        let options = TextProofreaderOptions()
        options.baseOptions.modelAssetPath = modelPath
        if let maxTokens { options.maxTokens = maxTokens }
        let proofreader = try TextProofreader(options: options)
        let handle = withLock {
          let handle = self.nextHandle
          self.nextHandle += 1
          self.proofreaders[handle] = proofreader
          return handle
        }
        complete(result, value: handle)
      } catch {
        complete(result, error: error)
      }
    }
  }

  private func proofread(_ call: FlutterMethodCall, result: @escaping FlutterResult) throws {
    let arguments = try call.argumentsMap()
    let handle = try arguments.requiredInt64("handle")
    let text = try arguments.requiredString("text")
    guard let proofreader = withLock({ proofreaders[handle] }) else {
      throw PluginError.invalidArgument("Unknown TextProofreader handle: \(handle)")
    }
    worker.async { [weak self] in
      guard let self else { return }
      do {
        let output = try proofreader.proofread(text: text)
        complete(
          result,
          value: [
            "text": output.proofreadText,
            "corrections": output.corrections.map(correctionMap),
          ]
        )
      } catch {
        complete(result, error: error)
      }
    }
  }

  private func proofreadStreaming(
    _ call: FlutterMethodCall,
    result: @escaping FlutterResult
  ) throws {
    try requireEventSink()
    let arguments = try call.argumentsMap()
    let handle = try arguments.requiredInt64("handle")
    let text = try arguments.requiredString("text")
    let requestId = try arguments.requiredString("requestId")
    guard let proofreader = withLock({ proofreaders[handle] }) else {
      throw PluginError.invalidArgument("Unknown TextProofreader handle: \(handle)")
    }
    worker.async { [weak self] in
      guard let self else { return }
      do {
        try proofreader.proofreadStreaming(text: text) { [weak self] output, error in
          guard let self else { return }
          if let error {
            emitError(requestId: requestId, error: error)
            return
          }
          guard let output else { return }
          emit([
            "kind": "data",
            "requestId": requestId,
            "text": output.chunk,
            "isDone": output.done,
            "corrections": output.corrections?.map(correctionMap) as Any,
          ])
          if output.done { emitDone(requestId: requestId) }
        }
        complete(result, value: nil)
      } catch {
        complete(result, error: error)
      }
    }
  }

  private func closeProofreader(_ call: FlutterMethodCall, result: @escaping FlutterResult) throws {
    let handle = try call.argumentsMap().requiredInt64("handle")
    guard let proofreader = withLock({ proofreaders.removeValue(forKey: handle) }) else {
      throw PluginError.invalidArgument("Unknown TextProofreader handle: \(handle)")
    }
    worker.async { [weak self] in
      guard let self else { return }
      do {
        try proofreader.close()
        complete(result, value: nil)
      } catch {
        complete(result, error: error)
      }
    }
  }

  private func createSummarizer(_ call: FlutterMethodCall, result: @escaping FlutterResult) throws {
    let arguments = try call.argumentsMap()
    let modelPath = try arguments.requiredString("modelPath")
    let maxTokens = arguments.optionalInt("maxTokens")
    let mode: TextSummarizerMode
    switch try arguments.requiredString("mode") {
    case "tldr": mode = .tldr
    case "keyPoints": mode = .keyPoints
    default: throw PluginError.invalidArgument("Unknown TextSummarizer mode")
    }
    worker.async { [weak self] in
      guard let self else { return }
      do {
        let options = TextSummarizerOptions()
        options.baseOptions.modelAssetPath = modelPath
        options.mode = mode
        if let maxTokens { options.maxTokens = maxTokens }
        let summarizer = try TextSummarizer(options: options)
        let handle = withLock {
          let handle = self.nextHandle
          self.nextHandle += 1
          self.summarizers[handle] = summarizer
          return handle
        }
        complete(result, value: handle)
      } catch {
        complete(result, error: error)
      }
    }
  }

  private func summarize(_ call: FlutterMethodCall, result: @escaping FlutterResult) throws {
    let arguments = try call.argumentsMap()
    let handle = try arguments.requiredInt64("handle")
    let text = try arguments.requiredString("text")
    guard let summarizer = withLock({ summarizers[handle] }) else {
      throw PluginError.invalidArgument("Unknown TextSummarizer handle: \(handle)")
    }
    worker.async { [weak self] in
      guard let self else { return }
      do {
        complete(result, value: ["summary": try summarizer.summarize(text: text).summary])
      } catch {
        complete(result, error: error)
      }
    }
  }

  private func summarizeStreaming(
    _ call: FlutterMethodCall,
    result: @escaping FlutterResult
  ) throws {
    try requireEventSink()
    let arguments = try call.argumentsMap()
    let handle = try arguments.requiredInt64("handle")
    let text = try arguments.requiredString("text")
    let requestId = try arguments.requiredString("requestId")
    guard let summarizer = withLock({ summarizers[handle] }) else {
      throw PluginError.invalidArgument("Unknown TextSummarizer handle: \(handle)")
    }
    worker.async { [weak self] in
      guard let self else { return }
      do {
        try summarizer.summarizeStreaming(text: text) { [weak self] output, error in
          guard let self else { return }
          if let error {
            emitError(requestId: requestId, error: error)
            return
          }
          guard let output else { return }
          emit([
            "kind": "data",
            "requestId": requestId,
            "text": output.chunk,
            "isDone": output.done,
          ])
          if output.done { emitDone(requestId: requestId) }
        }
        complete(result, value: nil)
      } catch {
        complete(result, error: error)
      }
    }
  }

  private func closeSummarizer(_ call: FlutterMethodCall, result: @escaping FlutterResult) throws {
    let handle = try call.argumentsMap().requiredInt64("handle")
    guard let summarizer = withLock({ summarizers.removeValue(forKey: handle) }) else {
      throw PluginError.invalidArgument("Unknown TextSummarizer handle: \(handle)")
    }
    worker.async { [weak self] in
      guard let self else { return }
      do {
        try summarizer.close()
        complete(result, value: nil)
      } catch {
        complete(result, error: error)
      }
    }
  }

  public func onListen(
    withArguments arguments: Any?,
    eventSink events: @escaping FlutterEventSink
  ) -> FlutterError? {
    withLock { eventSink = events }
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    withLock { eventSink = nil }
    return nil
  }

  public func detachFromEngine(for registrar: FlutterPluginRegistrar) {
    methodChannel?.setMethodCallHandler(nil)
    eventChannel?.setStreamHandler(nil)
    let openTasks = withLock { () -> ([TextProofreader], [TextSummarizer]) in
      let tasks = (Array(proofreaders.values), Array(summarizers.values))
      proofreaders.removeAll()
      summarizers.removeAll()
      eventSink = nil
      return tasks
    }
    worker.async {
      for task in openTasks.0 { try? task.close() }
      for task in openTasks.1 { try? task.close() }
    }
  }

  private func requireEventSink() throws {
    guard withLock({ eventSink != nil }) else {
      throw PluginError.invalidArgument(
        "Listen to the mp_text event channel before starting a stream")
    }
  }

  private func emit(_ event: [String: Any]) {
    DispatchQueue.main.async { [weak self] in
      guard let self else { return }
      let sink = withLock { self.eventSink }
      sink?(event)
    }
  }

  private func emitError(requestId: String, error: Error) {
    emit([
      "kind": "error",
      "requestId": requestId,
      "code": "internal",
      "message": error.localizedDescription,
    ])
  }

  private func emitDone(requestId: String) {
    emit(["kind": "done", "requestId": requestId])
  }

  private func complete(_ result: @escaping FlutterResult, value: Any?) {
    DispatchQueue.main.async { result(value) }
  }

  private func complete(_ result: @escaping FlutterResult, error: Error) {
    DispatchQueue.main.async { result(self.flutterError(error, code: "internal")) }
  }

  private func flutterError(_ error: Error, code: String) -> FlutterError {
    FlutterError(code: code, message: error.localizedDescription, details: nil)
  }

  private func correctionMap(_ correction: Correction) -> [String: String] {
    let type: String
    switch correction.type {
    case .same: type = "same"
    case .insertion: type = "insertion"
    case .deletion: type = "deletion"
    @unknown default: type = "unknown"
    }
    return ["type": type, "text": correction.text]
  }

  private func withLock<T>(_ operation: () -> T) -> T {
    lock.lock()
    defer { lock.unlock() }
    return operation()
  }
}

private enum PluginError: LocalizedError {
  case invalidArgument(String)

  var errorDescription: String? {
    switch self {
    case .invalidArgument(let message): return message
    }
  }
}

extension FlutterMethodCall {
  fileprivate func argumentsMap() throws -> [String: Any] {
    guard let value = arguments as? [String: Any] else {
      throw PluginError.invalidArgument("Arguments must be a map")
    }
    return value
  }
}

extension Dictionary where Key == String, Value == Any {
  fileprivate func requiredString(_ key: String) throws -> String {
    guard let value = self[key] as? String, !value.isEmpty else {
      throw PluginError.invalidArgument("\(key) must be a non-empty string")
    }
    return value
  }

  fileprivate func requiredInt64(_ key: String) throws -> Int64 {
    guard let value = self[key] as? NSNumber else {
      throw PluginError.invalidArgument("\(key) must be an integer")
    }
    return value.int64Value
  }

  fileprivate func optionalInt(_ key: String) -> Int? {
    (self[key] as? NSNumber)?.intValue
  }
}

File: packages/mp_text/android/src/main/kotlin/dev/ifiokjr/mptext/MpTextPlugin.kt
package dev.ifiokjr.mptext

import android.content.Context
import android.os.Handler
import android.os.Looper
import com.google.mediapipe.tasks.text.textproofreader.TextProofreader
import com.google.mediapipe.tasks.text.textproofreader.TextProofreaderResult
import com.google.mediapipe.tasks.text.textproofreader.TextProofreaderStreamingResult
import com.google.mediapipe.tasks.text.textsummarizer.TextSummarizer
import com.google.mediapipe.tasks.text.textsummarizer.TextSummarizerStreamingResult
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicLong

/** Android implementation of the mobile-only MediaPipe text generation tasks. */
class MpTextPlugin :
    FlutterPlugin,
    MethodChannel.MethodCallHandler,
    EventChannel.StreamHandler {
    private lateinit var applicationContext: Context
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private val mainHandler = Handler(Looper.getMainLooper())
    private val executor: ExecutorService = Executors.newCachedThreadPool()
    private val nextHandle = AtomicLong(1)
    private val proofreaders = ConcurrentHashMap<Long, TextProofreader>()
    private val summarizers = ConcurrentHashMap<Long, TextSummarizer>()

    @Volatile
    private var eventSink: EventChannel.EventSink? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = binding.applicationContext
        methodChannel = MethodChannel(binding.binaryMessenger, METHOD_CHANNEL)
        eventChannel = EventChannel(binding.binaryMessenger, EVENT_CHANNEL)
        methodChannel.setMethodCallHandler(this)
        eventChannel.setStreamHandler(this)
    }

    override fun onMethodCall(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        try {
            when (call.method) {
                "proofreader.create" -> createProofreader(call, result)
                "proofreader.proofread" -> proofread(call, result)
                "proofreader.stream" -> proofreadStreaming(call, result)
                "proofreader.close" -> closeProofreader(call, result)
                "summarizer.create" -> createSummarizer(call, result)
                "summarizer.summarize" -> summarize(call, result)
                "summarizer.stream" -> summarizeStreaming(call, result)
                "summarizer.close" -> closeSummarizer(call, result)
                else -> result.notImplemented()
            }
        } catch (error: IllegalArgumentException) {
            result.error("invalid_argument", error.message, null)
        }
    }

    private fun createProofreader(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        val modelPath = call.requiredString("modelPath")
        val maxTokens = call.optionalInt("maxTokens")
        executor.execute {
            try {
                val options = TextProofreader.TextProofreaderOptions.builder().setModelPath(modelPath)
                maxTokens?.let(options::setMaxNumTokens)
                val proofreader = TextProofreader.createFromOptions(applicationContext, options.build())
                val handle = nextHandle.getAndIncrement()
                proofreaders[handle] = proofreader
                result.successOnMain(handle)
            } catch (error: Throwable) {
                result.errorOnMain(error)
            }
        }
    }

    private fun proofread(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        val handle = call.requiredLong("handle")
        val text = call.requiredString("text")
        val proofreader =
            proofreaders[handle]
                ?: throw IllegalArgumentException("Unknown TextProofreader handle: $handle")
        executor.execute {
            try {
                val output = proofreader.proofread(text)
                result.successOnMain(
                    mapOf(
                        "text" to output.proofreadText,
                        "corrections" to output.corrections.toDartCorrections(),
                    ),
                )
            } catch (error: Throwable) {
                result.errorOnMain(error)
            }
        }
    }

    private fun proofreadStreaming(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        requireEventSink()
        val handle = call.requiredLong("handle")
        val text = call.requiredString("text")
        val requestId = call.requiredString("requestId")
        val proofreader =
            proofreaders[handle]
                ?: throw IllegalArgumentException("Unknown TextProofreader handle: $handle")
        executor.execute {
            try {
                proofreader.proofreadStreaming(
                    text,
                    object : TextProofreader.ProofreaderResultCallback {
                        override fun onNext(value: TextProofreaderStreamingResult) {
                            emit(
                                mapOf(
                                    "kind" to "data",
                                    "requestId" to requestId,
                                    "text" to value.chunk,
                                    "isDone" to value.isDone,
                                    "corrections" to value.corrections?.toDartCorrections(),
                                ),
                            )
                        }

                        override fun onError(error: Throwable) {
                            emitError(requestId, error)
                        }

                        override fun onDone() {
                            emitDone(requestId)
                        }
                    },
                )
                result.successOnMain(null)
            } catch (error: Throwable) {
                result.errorOnMain(error)
            }
        }
    }

    private fun closeProofreader(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        val handle = call.requiredLong("handle")
        val proofreader =
            proofreaders.remove(handle)
                ?: throw IllegalArgumentException("Unknown TextProofreader handle: $handle")
        executor.execute {
            try {
                proofreader.close()
                result.successOnMain(null)
            } catch (error: Throwable) {
                result.errorOnMain(error)
            }
        }
    }

    private fun createSummarizer(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        val modelPath = call.requiredString("modelPath")
        val maxTokens = call.optionalInt("maxTokens")
        val mode =
            when (call.requiredString("mode")) {
                "tldr" -> TextSummarizer.TextSummarizerOptions.Mode.TLDR
                "keyPoints" -> TextSummarizer.TextSummarizerOptions.Mode.KEYPOINTS
                else -> throw IllegalArgumentException("Unknown TextSummarizer mode")
            }
        executor.execute {
            try {
                val options =
                    TextSummarizer.TextSummarizerOptions
                        .builder()
                        .setModelPath(modelPath)
                        .setMode(mode)
                maxTokens?.let(options::setMaxNumTokens)
                val summarizer = TextSummarizer.createFromOptions(applicationContext, options.build())
                val handle = nextHandle.getAndIncrement()
                summarizers[handle] = summarizer
                result.successOnMain(handle)
            } catch (error: Throwable) {
                result.errorOnMain(error)
            }
        }
    }

    private fun summarize(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        val handle = call.requiredLong("handle")
        val text = call.requiredString("text")
        val summarizer =
            summarizers[handle]
                ?: throw IllegalArgumentException("Unknown TextSummarizer handle: $handle")
        executor.execute {
            try {
                result.successOnMain(mapOf("summary" to summarizer.summarize(text).summary))
            } catch (error: Throwable) {
                result.errorOnMain(error)
            }
        }
    }

    private fun summarizeStreaming(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        requireEventSink()
        val handle = call.requiredLong("handle")
        val text = call.requiredString("text")
        val requestId = call.requiredString("requestId")
        val summarizer =
            summarizers[handle]
                ?: throw IllegalArgumentException("Unknown TextSummarizer handle: $handle")
        executor.execute {
            try {
                summarizer.summarizeStreaming(
                    text,
                    object : TextSummarizer.SummarizationResultCallback {
                        override fun onNext(value: TextSummarizerStreamingResult) {
                            emit(
                                mapOf(
                                    "kind" to "data",
                                    "requestId" to requestId,
                                    "text" to value.chunk,
                                    "isDone" to value.isDone,
                                ),
                            )
                        }

                        override fun onError(error: Throwable) {
                            emitError(requestId, error)
                        }

                        override fun onDone() {
                            emitDone(requestId)
                        }
                    },
                )
                result.successOnMain(null)
            } catch (error: Throwable) {
                result.errorOnMain(error)
            }
        }
    }

    private fun closeSummarizer(
        call: MethodCall,
        result: MethodChannel.Result,
    ) {
        val handle = call.requiredLong("handle")
        val summarizer =
            summarizers.remove(handle)
                ?: throw IllegalArgumentException("Unknown TextSummarizer handle: $handle")
        executor.execute {
            try {
                summarizer.close()
                result.successOnMain(null)
            } catch (error: Throwable) {
                result.errorOnMain(error)
            }
        }
    }

    private fun requireEventSink() {
        requireNotNull(eventSink) { "Listen to the mp_text event channel before starting a stream" }
    }

    private fun emit(event: Map<String, Any?>) {
        mainHandler.post { eventSink?.success(event) }
    }

    private fun emitError(
        requestId: String,
        error: Throwable,
    ) {
        emit(
            mapOf(
                "kind" to "error",
                "requestId" to requestId,
                "code" to "internal",
                "message" to error.safeMessage(),
            ),
        )
    }

    private fun emitDone(requestId: String) {
        emit(mapOf("kind" to "done", "requestId" to requestId))
    }

    override fun onListen(
        arguments: Any?,
        events: EventChannel.EventSink,
    ) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        eventSink = null
        val openProofreaders = proofreaders.values.toList()
        val openSummarizers = summarizers.values.toList()
        proofreaders.clear()
        summarizers.clear()
        executor.execute {
            openProofreaders.forEach { runCatching { it.close() } }
            openSummarizers.forEach { runCatching { it.close() } }
        }
        executor.shutdown()
    }

    private fun MethodCall.requiredString(name: String): String =
        argument<String>(name)?.takeIf { it.isNotEmpty() }
            ?: throw IllegalArgumentException("$name must be a non-empty string")

    private fun MethodCall.requiredLong(name: String): Long =
        argument<Number>(name)?.toLong()
            ?: throw IllegalArgumentException("$name must be an integer")

    private fun MethodCall.optionalInt(name: String): Int? = argument<Number>(name)?.toInt()

    private fun MethodChannel.Result.successOnMain(value: Any?) {
        mainHandler.post { success(value) }
    }

    private fun MethodChannel.Result.errorOnMain(error: Throwable) {
        mainHandler.post { error("internal", error.safeMessage(), null) }
    }

    private fun Throwable.safeMessage(): String = message ?: javaClass.simpleName

    private fun List<TextProofreaderResult.Correction>.toDartCorrections(): List<Map<String, String>> =
        map { correction ->
            mapOf(
                "type" to correction.type.name.lowercase(),
                "text" to correction.text,
            )
        }

    private companion object {
        const val METHOD_CHANNEL = "dev.ifiokjr.mp_text/methods"
        const val EVENT_CHANNEL = "dev.ifiokjr.mp_text/events"
    }
}
